README.md
# jubakit: Jubatus Toolkit

jubakit is a Python module to access Jubatus features easily.
jubakit can be used in conjunction with [scikit-learn](http://scikit-learn.org/) so that you can use powerful features like cross validation and model evaluation.

## Install

```
pip install jubakit
```

### Requirements

* Python 2.7, 3.3, 3.4 or 3.5.
* [Jubatus](http://jubat.us/en/quickstart.html) needs to be installed.
* Although not mandatory, [scikit-learn](http://scikit-learn.org/stable/install.html) is required to use some features like K-fold cross validation.

```
pip install numpy scipy scikit-learn
```

## Quick Start

```py
from jubakit.classifier import Classifier, Schema, Dataset, Config
from jubakit.loader.csv import CSVLoader

# Load a CSV file.
loader = CSVLoader('iris.csv')

# Define types for each column in the CSV file.
schema = Schema({
  'Species': Schema.LABEL,
  'Sepal.Length': Schema.NUMBER,
  'Sepal.Width': Schema.NUMBER,
  'Petal.Length': Schema.NUMBER,
  'Petal.Width': Schema.NUMBER,
})

# Get the shuffled dataset.
dataset = Dataset(loader, schema).shuffle()

# Run the classifier service (`jubaclassifier` process).
classifier = Classifier.run(Config())

# Train the classifier.
for _ in classifier.train(dataset): pass

# Classify using the trained classifier.
for (idx, label, result) in classifier.classify(dataset):
  print("true label: {0}, estimated label: {1}".format(label, result[0][0]))
```

## Examples by Task

See the `example` directory for working examples.

| Example                   | Topics                                        | Requires scikit-learn |
|---------------------------|-----------------------------------------------|-----------------------|
| classifier_csv.py         | Handling CSV file and numeric features        |                       |
| classifier_shogun.py      | Handling CSV file and string features         |                       |
| classifier_digits.py      | Handling toy dataset (digits)                 | ✓                     |
| classifier_libsvm.py      | Handling LIBSVM file                          | ✓                     |
| classifier_kfold.py       | K-fold cross validation and metrics           | ✓                     |
| classifier_parameter.py   | Finding best hyper parameter                  | ✓                     |
| classifier_bulk.py        | Bulk Train-Test Classifier                    |                       |

## Concepts

* **Loader** fetches data from various data sources (e.g., CSV file, RDBMS, MQ, Twitter stream, etc.) in key-value format.
* **Schema** defines the data type (string feature, numeric feature, ground truth (label), etc.) for each keys of data loaded by Loader.
* **Dataset** is an abstract representation of a sequence of data that binds Loader and Schema.
* **Service** receives Dataset and make update/analyze RPC call to Jubatus servers.
* **Metrics** can be used to evaluate a performance (precision) of a model.


ChangeLog.rst
ChangeLog
====================================================

Release 0.6.2 - 2019/01/28
--------------------------

* New Features
    * Add Bandit service (#129, #131)
    * Add support for clear_row method of Recommender (#136, #137)

* Improvements
    * Add visualize example using TensorBoard (#131)

* Bug Fixes
    * Fix schema not working with service specialized column definition (#134, #135)
    * Fix Twitter Loader not working with latest tweepy (#138)

Release 0.6.1 - 2018/10/29
--------------------------

* New Features
    * Add postgresql loader (#125)
    * Add support for filtering API of Recommender (#126, #127)
* Bug Fixes
    * Fix ``classifier_kfold.py`` example not working with sklearn 0.20 or later (#128)

Release 0.6.0 (2018-08-27)
--------------------------

* New Features
    * Add Burst service (#94, #124)

Release 0.5.5 (2018-04-23)
---------------------------------------

* New Features
    * Add support for distance option of Clustering (#123)

* Documentation
    * Translate documents into Japanese (#104, #105, #108, #110, #111, #113-122)

Release 0.5.4 (2018-02-26)
---------------------------------------

* New Features
    * Add Nearest Neighbor service (#92, #103)
    * Add support for add_bulk of Anomaly (#106)

Release 0.5.3 (2017-12-18)
---------------------------------------

* New Features
    * Add Clustering service (#93, #98)

Release 0.5.2 (2017-10-30)
---------------------------------------

* Changes to Supported Systems
    * Deprecate  Python 2.6 support (#99, #100)


Release 0.5.1 (2017-08-28)
---------------------------------------

* New Features
    * Add ``--transform`` option to ``jubamodel`` command (#75, #83)

* Improvements
    * Fix library version for Python 2.6 (#88, #91)

Release 0.5.0 (2017-04-24)
---------------------------------------

* New Features
    * Add Regression service (#82, #87)
    * Add ``--replace-config`` / ``--replace-version`` options to ``jubamodel`` command (#76, #77)

* Improvements
    * Improve color of version number in docs (#80, #81)

* Bug Fixes
    * Fix CSVLoader not working with ``restkey`` option in Python 2 (#78, #79)

Release 0.4.2 (2017-02-27)
---------------------------------------

* New Features
    * Add scikit-learn wrapper (#73)

* Bug Fixes
    * Fix dead links in documents (#74)

Release 0.4.1 (2016-12-26)
---------------------------------------

* New Features
    * Support Embedded Jubatus (#70)

* Improvements
    * Implement ``__repr__`` method to base classes (#42, #71)
    * Add ``jubakit.model.JubaModel`` example (#68)
    * Add ``jubakit.model.JubaDump`` example (#69, #71)

* Bug Fixes
    * Fix document typos (#67)

Release 0.4.0 (2016-10-31)
---------------------------------------

* New Features
    * Add Recommender service (#52, #58)
    * Add model file manipulation tools (#4, #62)
    * Add ``ConcatLoader`` (#61)

* Improvements
    * Support Jubatus 1.0 (#66)
    * Improve Classifier Schema to accept columns without LABEL (#40, #57)
    * Improve handlign of bool-typed values (#9, #63)
    * Add option to apply Softmax to the resulting score in Classifier (#59)
    * Implement feature to enable logging by environment variable (#39, #65)
    * Improve RPC port allocation method (#44, #54)
    * Improve jubakit to work with non-standard Homebrew installations on macOS (#32, #35)
    * Update theme of docs (#60)

* Bug Fixes
    * Fix non-static Datasets does not allow automatic Schema prediction (#41, #64)
    * Fix error handling in ``jubash`` command (#55)
    * Fix error message when Jubatus servers are not installed on macOS (#56)

Release 0.3.0 (2016-08-29)
---------------------------------------

* New Features
    * Add Shell feature and ``jubash`` command (#38, #45, #51)

* Improvements
    * Add API reference and design documents (#33, #36, #43, #53)
    * Add ``from_data`` method to Classifier service (#46)

* Bug Fixes
    * Fix CSVLoader parameter handling (#34)
    * Fix typo in error message format (#49, #50)

Release 0.2.2 (2016-07-25)
---------------------------------------

* Improvements
    * Add hyper-parameter tuning example using hyperopt library (#28)
    * Fix warnings when using ``numpy`` record (#29)

* Bug Fixes
    * Fix log output from server not captured correctly (#30, #31)

Release 0.2.1 (2016-06-27)
---------------------------------------

* Improvements
    * Add support for ``cosine`` and ``euclidean`` method of Classifier (#27)
    * Improve default configuration of ``NN`` method of Classifier to use multiple CPU cores (#26)
    * Improve Anomaly service to validate invalid method name (#25)
    * Fix deprecated test warning in Python 3.5 (#24)

Release 0.2.0 (2016-05-30)
---------------------------------------

* New Features
    * Add support for Anomaly engine (#20)
    * Add support for Weight engine (#19)

* Improvements
    * Add logging system (#12, #17)
    * Add ``get_status`` API to Service (#15)
    * Add seed option to ``shuffle`` API to Dataset (#14)
    * Revise ``Loader`` API (#13)
    * ``Config.add_mecab`` API now accepts feature filters as list (#16)
    * Improve documentation (#11)
    * Add tests (#21)

* Bug Fixes
    * Fix missing records not ignored as expected (#22)
    * Fix ``get_label`` of Classifier service raise unexpected error when the dataset is not static (#21)
    * Fix ``LineBasedStreamLoader`` not closing file when iteration is terminated (#10)
    * Fix ``classifer_bulk.py`` example not using config object (#18)

Release 0.1.0 (2016-04-25)
---------------------------------------

Initial release.


jubakit/dumb.py
# -*- coding: utf-8 -*-

"""
*Dumb* Service is a kind of temporary implementations of Services.
They are defined just for convenience.

Unlike *Real* Services (Classifier, Anomaly, ...) which are defined
in each file (classifier.py, anomaly.py, ...), Dumb Services cannot
handle Datasets and Schemas.

Each service has a field called ``CONFIG``, which provides a default
config data structure for the service.  So you can use jubakit to start
a Jubatus server processe, then directly use the raw Client class to
make RPC calls.

  >>> from jubakit.dumb import Stat
  >>> service = Stat.run(Stat.CONFIG)
  >>> client = service._client()
  >>> client.push('x', 12)
"""

from __future__ import absolute_import, division, print_function, unicode_literals

from jubakit.base import BaseService

import jubatus
import jubatus.embedded

class Bandit(BaseService):
  CONFIG = {'method': 'ucb1', 'parameter': {'assume_unrewarded': False}}

  @classmethod
  def name(cls):           return 'bandit'
  @classmethod
  def _client_class(cls):  return jubatus.bandit.client.Bandit
  @classmethod
  def _embedded_class(cls):  return jubatus.embedded.Bandit

class Burst(BaseService):
  CONFIG = {'method': 'burst', 'parameter': {'result_window_rotate_size': 5, 'max_reuse_batch_num': 5, 'batch_interval': 10, 'window_batch_size': 5, 'costcut_threshold': -1}}

  @classmethod
  def name(cls):           return 'burst'
  @classmethod
  def _client_class(cls):  return jubatus.burst.client.Burst
  @classmethod
  def _embedded_class(cls):  return jubatus.embedded.Burst

class Clustering(BaseService):
  CONFIG = {'method': 'kmeans', 'parameter': {'k': 3, 'seed': 0}, 'compressor_method': 'simple', 'compressor_parameter': {'bucket_size': 1000}, 'distance': 'euclidean', 'converter': {'string_types': {'bigram': {'method': 'ngram', 'char_num': '2'}, 'trigram': {'method': 'ngram', 'char_num': '3'}, 'unigram': {'method': 'ngram', 'char_num': '1'}}, 'num_filter_types': {}, 'num_rules': [{'type': 'num', 'key': '*'}], 'num_filter_rules': [], 'string_filter_rules': [], 'num_types': {}, 'string_filter_types': {}, 'string_rules': [{'sample_weight': 'tf', 'global_weight': 'idf', 'type': 'bigram', 'key': '*'}]}}

  @classmethod
  def name(cls):           return 'clustering'
  @classmethod
  def _client_class(cls):  return jubatus.clustering.client.Clustering
  @classmethod
  def _embedded_class(cls):  return jubatus.embedded.Clustering

class Graph(BaseService):
  CONFIG = {'method': 'graph_wo_index', 'parameter': {'damping_factor': 0.9, 'landmark_num': 5}}

  @classmethod
  def name(cls):           return 'graph'
  @classmethod
  def _client_class(cls):  return jubatus.graph.client.Graph
  @classmethod
  def _embedded_class(cls):  return jubatus.embedded.Graph

class NearestNeighbor(BaseService):
  CONFIG = {'method': 'lsh', 'parameter': {'hash_num': 64}, 'converter': {'string_types': {'bigram': {'method': 'ngram', 'char_num': '2'}, 'trigram': {'method': 'ngram', 'char_num': '3'}, 'unigram': {'method': 'ngram', 'char_num': '1'}}, 'num_filter_types': {}, 'num_rules': [{'type': 'num', 'key': '*'}], 'num_filter_rules': [], 'string_filter_rules': [], 'num_types': {}, 'string_filter_types': {}, 'string_rules': [{'sample_weight': 'tf', 'global_weight': 'idf', 'type': 'bigram', 'key': '*'}]}}

  @classmethod
  def name(cls):           return 'nearest_neighbor'
  @classmethod
  def _client_class(cls):  return jubatus.nearest_neighbor.client.NearestNeighbor
  @classmethod
  def _embedded_class(cls):  return jubatus.embedded.NearestNeighbor

class Recommender(BaseService):
  CONFIG = {'method': 'inverted_index', 'converter': {'string_types': {'bigram': {'method': 'ngram', 'char_num': '2'}, 'trigram': {'method': 'ngram', 'char_num': '3'}, 'unigram': {'method': 'ngram', 'char_num': '1'}}, 'num_filter_types': {}, 'num_rules': [{'type': 'num', 'key': '*'}], 'num_filter_rules': [], 'string_filter_rules': [], 'num_types': {}, 'string_filter_types': {}, 'string_rules': [{'sample_weight': 'tf', 'global_weight': 'idf', 'type': 'bigram', 'key': '*'}]}}

  @classmethod
  def name(cls):           return 'recommender'
  @classmethod
  def _client_class(cls):  return jubatus.recommender.client.Recommender
  @classmethod
  def _embedded_class(cls):  return jubatus.embedded.Recommender

class Regression(BaseService):
  CONFIG = {'method': 'PA1', 'parameter': {'sensitivity': 0.1, 'regularization_weight': 3.402823e+38}, 'converter': {'string_types': {'bigram': {'method': 'ngram', 'char_num': '2'}, 'trigram': {'method': 'ngram', 'char_num': '3'}, 'unigram': {'method': 'ngram', 'char_num': '1'}}, 'num_filter_types': {}, 'num_rules': [{'type': 'num', 'key': '*'}], 'num_filter_rules': [], 'string_filter_rules': [], 'num_types': {}, 'string_filter_types': {}, 'string_rules': [{'sample_weight': 'tf', 'global_weight': 'idf', 'type': 'bigram', 'key': '*'}]}}

  @classmethod
  def name(cls):           return 'regression'
  @classmethod
  def _client_class(cls):  return jubatus.regression.client.Regression
  @classmethod
  def _embedded_class(cls):  return jubatus.embedded.Regression

class Stat(BaseService):
  CONFIG = {'window_size': 128}

  @classmethod
  def name(cls):           return 'stat'
  @classmethod
  def _client_class(cls):  return jubatus.stat.client.Stat
  @classmethod
  def _embedded_class(cls):  return jubatus.embedded.Stat
